### ISSUE TITLE
druntime issues on Alpine Linux (Musl C runtime)

### ISSUE THREAD

**Issue by kirbyfan64** [ldc-developers/ldc#2341]
Ok, so I've been trying to compile LDC on Alpine Linux, and it's been...interesting...

So, first off: because #1664 was only merged into the "modern" LDC tree, it's *impossible* to bootstrap LDC 0.17.5. I had to manually apply the patch onto the LDC source, and then I could bootstrap.

However, on Alpine, the `backtrace` functions are in [libexecinfo](http://www.freshports.org/devel/libexecinfo), not libc:

```
/usr/local/lib/libdruntime-ldc.a(dwarf.o): In function `_D2rt9backtrace5dwarf23traceHandlerOpApplyImplFAxPvMDFKmKxAaZiZi':
rt/backtrace/dwarf.d:(.text._D2rt9backtrace5dwarf23traceHandlerOpApplyImplFAxPvMDFKmKxAaZiZi[_D2rt9backtrace5dwarf23traceHandlerOpApplyImplFAxPvMDFKmKxAaZiZi]+0x2a): undefined reference to `backtrace_symbols'
/usr/local/lib/libdruntime-ldc.a(runtime.o): In function `_D4core7runtime19defaultTraceHandlerFPvZC6object9Throwable9TraceInfo':
core/runtime.d:(.text._D4core7runtime19defaultTraceHandlerFPvZC6object9Throwable9TraceInfo[_D4core7runtime19defaultTraceHandlerFPvZC6object9Throwable9TraceInfo]+0x6b): undefined reference to `backtrace'
/usr/local/lib/libdruntime-ldc.a(runtime.o): In function `_D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPS4core3sys5posix6signal9siginfo_tPvZv':
core/runtime.d:(.text._D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPS4core3sys5posix6signal9siginfo_tPvZv[_D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPS4core3sys5posix6signal9siginfo_tPvZv]+0x23): undefined reference to `backtrace'
core/runtime.d:(.text._D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPS4core3sys5posix6signal9siginfo_tPvZv[_D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPS4core3sys5posix6signal9siginfo_tPvZv]+0x32): undefined reference to `backtrace_symbols_fd'
/usr/local/lib/libdruntime-ldc.a(runtime.o): In function `_D4core7runtime19defaultTraceHandlerFPvZ16DefaultTraceInfo6__ctorMFZC4core7runtime19defaultTraceHandlerFPvZ16DefaultTraceInfo':
core/runtime.d:(.text._D4core7runtime19defaultTraceHandlerFPvZ16DefaultTraceInfo6__ctorMFZC4core7runtime19defaultTraceHandlerFPvZ16DefaultTraceInfo[_D4core7runtime19defaultTraceHandlerFPvZ16DefaultTraceInfo6__ctorMFZC4core7runtime19defaultTraceHandlerFPvZ16DefaultTraceInfo]+0x11): undefined reference to `backtrace'
```

So `-lexecinfo` needs to be passed. Since I couldn't make *that* work, I gave up and hacked it in:

```
mkdir exec
cd exec
ar x /usr/lib/libexecinfo.a
cd ..
ar rcs lib/libdruntime-ldc.a exec/*
```

Now I'm stuck on exceptions not working, like so:

```
DevPC-LX2:~$ ldmd2 hello.d
DevPC-LX2:~$ ./hello
Segmentation fault (core dumped)
DevPC-LX2:~$ gdb hello
GNU gdb (GDB) 7.12.1
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-alpine-linux-musl".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from hello...done.
(gdb) r
Starting program: /home/user/hello 

Program received signal SIGSEGV, Segmentation fault.
0x00005555555b467d in getframeaddr ()
(gdb) bt
#0  0x00005555555b467d in getframeaddr ()
#1  0x00005555555a7c59 in backtrace ()
#2  0x00005555555b773f in core.runtime.defaultTraceHandler(void*) ()
#3  0x00005555555988d3 in _d_throw_exception ()
#4  0x000055555556f64f in D main ()
(gdb) 
```

where these are the contents of `hello.d`:

```d
import std.stdio;

void main() {
  throw new Exception("Hello, world!");
}
```

I'm thinking maybe it's related to [this issue](https://github.com/crystal-lang/crystal/issues/4276), and...now I'm stuck...

**Comment by kirbyfan64** [comment]
The following script should reproduce all the issues:

```
sudo apk add cmake extra-cmake-modules@testing ninja
sudo apk add curl-dev sqlite-dev llvm4-dev llvm4-static clang libexecinfo-dev

BASE=https://github.com/ldc-developers/ldc/releases/download

export CC=clang
export CXX=clang++

curl -Lo ldc0.tgz $BASE/v0.17.5/ldc-0.17.5-src.tar.gz
tar xvf ldc0.tgz
cd ldc-0*

curl https://patch-diff.githubusercontent.com/raw/ldc-developers/ldc/pull/1664.diff | \
  patch -p1

mkdir build
cd build
cmake .. -G Ninja -DCMAKE_BUILD_TYPE=Release
ninja

mkdir exec
cd exec
ar x /usr/lib/libexecinfo.a
cd ..
ar rcs lib/libdruntime-ldc.a exec/*

cd ../..

curl -Lo ldc.tgz $BASE/v1.4.0/ldc-1.4.0-src.tar.gz
tar xvf ldc.tgz
cd ldc-1*

mkdir build
cd build
PATH=$PATH:/home/user/ldc-0.17.5-src/build/bin \
  cmake .. -G Ninja -DCMAKE_BUILD_TYPE=Release
ninja

mkdir exec
cd exec
ar x /usr/lib/libexecinfo.a
cd ..
ar rcs lib/libdruntime-ldc.a exec/*

sudo ninja install
cd ../..
```

**Comment by JohanEngelen** [comment]
> -lexecinfo

To pass that, you can edit `ldc.conf` and add it to the default flags. `-L-lexecinfo` should work.

**Comment by JohanEngelen** [comment]
> So, first off: because #1664 was only merged into the "modern" LDC tree, it's impossible to bootstrap LDC 0.17.5. I had to manually apply the patch onto the LDC source, and then I could bootstrap.

Don't be afraid to submit PRs ;-)  See: https://github.com/ldc-developers/ldc/pull/2343

**Comment by kinke** [comment]
> To pass that, you can edit ldc.conf and add it to the default flags. `-L-lexecinfo` should work.

The same goes for PIC - you can add `-relocation-model=pic` to `etc/ldc2.conf` as well.

As to the crash when generating the stacktrace for the exception, you can temporarily disable that to move forward, see https://github.com/ldc-developers/druntime/pull/86/files#diff-a7ad725ec7185d73200e281adb599c04L494.

**Comment by Timbus** [comment]
Just chiming in: This is something I was interested in a few weeks ago. Had the exact same problem. The only difference is I bootstrapped LDC by using the frolvad/alpine-glibc docker image + packing libexecinfo into the prebuilt druntime, which works 'well enough' for bootstrapping. I assume it causes the CRuntime_Glibc flag to be set though, which forced me to comment out a use of `qsort_r`.

After seeing the last post by @kinke I can confirm that taking out the stack trace gets things working for both dynamic and statically linked binaries.

As an additional test I tried using libunwind in place of libexecinfo (as some people in the linked issues found it works better), but it appears to be missing `backtrace_symbols` as well as `backtrace_symbols_fd`. Are they actually used? Perhaps some `version(CRuntime_Musl)` blocks could be used to fix this..?

For now everything works(?) but I think the process could be improved.. Would be really cool to have LDC easily building alpine containers + static D binaries for microservice and tooling development. I'm pretty tired of reading Go code :)

**Comment by kinke** [comment]
After #2273 (predefined version `CRuntime_Musl`), the remaining adaptations should happen in druntime (I counted 138 occurrences of `CRuntime_Glibc` and 115 of `CRuntime_Bionic` in druntime v2.076.1), incl. a `pragma(lib, "...")` for the extra library.

**Comment by joakim-noah** [comment]
Are you two trying to run microservices?  If so, chime in on [this forum thread](http://forum.dlang.org/thread/twtbrappuarwhyzlchrb@forum.dlang.org).

**Comment by joakim-noah** [comment]
The just released 1.8 comes with an Alpine build, which only requires the `llvm5` package to run, `musl-dev` and `gcc` to link D apps, and `libcurl` and `tzdata` for the `std.net.curl` and `std.datetime.timezone` modules to work.
